### PR TITLE
feat(auth): Signup function rewrite, adds new types for auth

### PR DESCRIPTION
**Pull request by helgabalashova**
#### Description of changes
`Auth.signUp` function rewrite. Adds new types for Auth and HubEvents. Work in progress.

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 1]
```diff
@@ -15,29 +15,343 @@ import {
...
+			AuthParameters: {
+				username,
+				password
+			},
+			ClientId: clientId,
+			ClientMetadata: clientMetaData
+		});
+
+		if (signUpResult.isSignUpComplete) {
+			this.signInAfterUserConfirmed(initiateAuthCommand, username);
```
Comment (AllanZhengYP) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:229]: Missing `await`? Same as other 2 branches.

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 2]
```diff
@@ -15,29 +15,343 @@ import {
...
+							this.handleAutoSignIn(username, password, clientId, autoSignInClientMetaData, result);
+						}
+						resolve(result)
+					}
+				}
+			)
+		})
+	}
+
+	private handleAutoSignIn<UserAttributeKey extends AuthUserAttributeKey>(
```
Comment (AllanZhengYP) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:210]: This should an async function.

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 3]
```diff
@@ -15,29 +15,343 @@ import {
...
+			this.handleCodeAutoSignIn(initiateAuthCommand, username);
+		}
+
+	}
+
+	private async signInAfterUserConfirmed<UserAttributeKey extends AuthUserAttributeKey>(
+		initiateAuthCommand: InitiateAuthCommand,
+		username: string,
+		hubListenerRemoveToken?: () => void,
+		autoSignInPollingIntervalId?: ReturnType<typeof setInterval>
```
Comment (AllanZhengYP) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:242]: This function should not include logics specific to `handleLinkAutoSignIn` or `handleCodeAutoSignIn`. I'd remove them and move the individual code for cleaning resources back to the 2 functions.

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 4]
```diff
@@ -15,29 +15,343 @@ import {
...
+			)
+
+			if (hubListenerRemoveToken) {
+				hubListenerRemoveToken();
+			}
+
+			if (autoSignInPollingIntervalId) {
+				clearInterval(autoSignInPollingIntervalId);
+				this._storage.removeItem('amplify_polling_started');
+			}
```
Comment (AllanZhengYP) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:281]: These can be done outside of this function.

File: packages/auth/src/types/AmazonCognitoProvider.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
+
+// common types
+
+import { OAuthOpts } from "../../lib-esm/types"
+
+export type AuthPluginOptions = Record<string, any>
+
+export type ClientMetaData = {
+	[key:string]: string
+}
```
Comment (AllanZhengYP) [inline packages/auth/src/types/AmazonCognitoProvider.ts:10]: Need to confirm but I think we use `Record<string, string>` type in other places in codebase(newer part). Same as the other places in this file.

File: packages/auth/src/types/hubEvents.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
+import { AuthError } from "../Errors";
```
Comment (AllanZhengYP) [inline packages/auth/src/types/hubEvents.ts:1]: Not sure why this file is using double quotes and lint is not complain about it.

**Review by AllanZhengYP — COMMENTED** [review]
This is a solid start! Thank you for putting this together!

File: packages/auth/src/Providers/AmazonCognitoProvider.ts (hunk 3 above)
Comment (helgabalashova) [reply packages/auth/src/Providers/AmazonCognitoProvider.ts]: I removed code logic from here, but I don't see how link logic can be removed. We only remove listener if sign in succeeded. With link verification, I can't know if sign in succeeded within handle function. That's why it is being removed within this function once we know user signed in.

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 7]
```diff
@@ -15,29 +15,320 @@ import {
...
+				this.handleAutoSignIn(username, password, clientId, autoSignInClientMetaData, result);
+			}
+			return result;
+		} catch (error) {
+			dispatchAuthEvent(
+				AuthHubEvent.SIGN_UP_FAILURE,
+				error,
+				`${username} failed to sign-up`
+			);
+			return error;
```
Comment (AllanZhengYP) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:186]: Throw error?

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 8]
```diff
@@ -15,29 +15,320 @@ import {
...
+					response: authSignInResult,
+					user: null
+				},
+				`${username} has signed in successfully`
+			);
+
+			if (autoSignInPollingIntervalId) {
+				clearInterval(autoSignInPollingIntervalId);
+				this._storage.removeItem('amplify_polling_started');
+			}
```
Comment (AllanZhengYP) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:256]: I don't think I can find where `'amplify_polling_started'` is set. This if block can be moved below line 277. 

Also a super nit: I notice `'amplify-auto-sign-in'` is using the kabab casing whereas this is snake casing.

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 9]
```diff
@@ -15,29 +15,320 @@ import {
...
+
+			this._storage.removeItem('amplify-auto-sign-in');
+		} catch (error) {
+			logger.error(error);
```
Comment (AllanZhengYP) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:260]: Is it intended to swallow the error?

File: package.json [hunk 10]
```diff
@@ -122,5 +121,8 @@
 	"jest": {
 		"resetMocks": true,
 		"verbose": true
+	},
+	"dependencies": {
+		"@aws-sdk/client-cognito-identity-provider": "^3.198.0"
```
[the remark this answers is not in the document]
Comment (haverchuck) [reply package.json:126]: Yes, we are using the signup command.

File: packages/auth/src/Errors.ts [hunk 11]
```diff
@@ -19,35 +19,29 @@ import { AuthErrorMessages } from './types/AuthErrorMessages';
...
-		// Hack for making the custom error class work when transpiled to es5
-		// TODO: Delete the following 2 lines after we change the build target to >= es2015
-		this.constructor = AuthError;
-		Object.setPrototypeOf(this, AuthError.prototype);
-
-		this.name = 'AuthError';
-		this.log = log || message;
-
-		logger.error(this.log);
+	underlyingException?: Error;
```
Comment (haverchuck) [inline packages/auth/src/Errors.ts:22]: These will probably be changed a bit based on current API proposal. No need to modify this PR right now, though.

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 12]
```diff
@@ -15,29 +15,320 @@ import {
...
+
 export class AmazonCognitoProvider implements AuthProvider {
 	static category = 'Auth';
 	static providerName = 'AmazonCognito';
 
-	private _config;
+	private _config: AuthOptions;
+	private client: CognitoIdentityProviderClient;
+	private _storage: any;
+	private autoSignInInitiated = false;
```
Comment (haverchuck) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:74]: `autoSignInInitiated` will most likely be moved to a state machine. It's ok to have this here for now.

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 13]
```diff
@@ -15,29 +15,320 @@ import {
...
 	}
-	signUp(): Promise<any> {
-		throw new Error('Method not implemented.');
+
+	async signUp<PluginOptions extends AuthPluginOptions = CognitoSignUpOptions>(req: SignUpRequest<CognitoUserAttributeKey, PluginOptions>): Promise<AuthSignUpResult<CognitoUserAttributeKey>> {
+		if (!this._config?.userPoolId) {
+			this.rejectNoUserPool();
+		}
+
+		const clientId = this._config.userPoolId ?? '';
```
Comment (haverchuck) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:87]: Can we create service for userpool SDK calls? We can then make the client id part of the service constructor, and don't have to handle it in each individual API.

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 14]
```diff
@@ -15,29 +15,343 @@ import {
...
+		if (!this._config?.userPoolId) {
+			this.rejectNoUserPool();
+		}
+
+		const clientId = this._config.userPoolId ?? "";
+		const username: string = req.username;
+		const password: string = req.password;
+
+		if (!username) {
+			this.rejectAuthError(authErrorMessages.emptyUsername);
```
[the remark this answers is not in the document]
Comment (haverchuck) [reply packages/auth/src/Providers/AmazonCognitoProvider.ts:92]: These will change when error handling proposal is ironed out.

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 15]
```diff
@@ -15,29 +15,320 @@ import {
...
+
+		if (!password) {
+			this.rejectAuthError(authErrorMessages.emptyPassword);
+		}
+
+		const signUpCommandInput: SignUpCommandInput = {
+			ClientId: clientId,
+			Username: username,
+			Password: password,
+		};
```
Comment (haverchuck) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:103]: Creating the command input can be moved to a service file.

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 16]
```diff
@@ -15,29 +15,320 @@ import {
...
+		if (req.options?.autoSignIn?.enabled) {
+			autoSignIn = req.options.autoSignIn;
+			this._storage.setItem('amplify-auto-sign-in', 'true');
+			autoSignInClientMetaData = autoSignIn.clientMetaData ?? {};
+		}
+
+		const signUpCommand = new SignUpCommand(signUpCommandInput);
+
+		try {
+			const signUpCommandOutput = await this.client.send(signUpCommand);
```
Comment (haverchuck) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:138]: Creating the command and sending to the client can be part of the userpool service.

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 17]
```diff
@@ -15,29 +15,320 @@ import {
...
+			dispatchAuthEvent(
+				AuthHubEvent.SIGN_UP_FAILURE,
+				error,
+				`${username} failed to sign-up`
+			);
+			return error;
+		}
+	}
+
+	private async handleAutoSignIn<UserAttributeKey extends AuthUserAttributeKey>(
```
Comment (haverchuck) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:190]: All of this autoSignIn stuff can be moved to the state machine once it's ready, but it's fine to have it here for now.

File: packages/auth/src/types/AmazonCognitoProvider.ts [hunk 18]
```diff
@@ -0,0 +1,181 @@
+
+// common types
+
+import { OAuthOpts } from '../../lib-esm/types';
+
+export type AuthPluginOptions = Record<string, any>;
```
Comment (haverchuck) [inline packages/auth/src/types/AmazonCognitoProvider.ts:6]: Can create another types  file at the category level? So things like AuthPluginOptions would exist there, but Cognito-specific types would remain here.

**Review by haverchuck — CHANGES_REQUESTED** [review]
Left some comments, with three over-arching themes:
1. We can create a UserPoolService that is called from the API to make SDK calls. That way we leave the door open for better tree shaking opportunities, cleaner separation of concerns, and the ability to do things like set the clientId only once.
2. Take a look at the error handling proposal in the API doc.
3. We can create a separate types file for category-level types (I think maybe we talked about this but I don't remember the outcome).

File: packages/auth/src/Auth.ts [hunk 19]
```diff
@@ -101,9 +101,10 @@ export class AuthClass {
 		this._pluggable = null;
 	}
 
-	signUp(): Promise<any> {
+	signUp<PluginOptions extends Record<string, any>>(req: SignUpRequest<string, PluginOptions>): Promise<any>;
+	signUp<PluginOptions extends AuthPluginOptions = CognitoSignUpOptions>(req: SignUpRequest<CognitoUserAttributeKey,PluginOptions>):Promise<AuthSignUpResult<CognitoUserAttributeKey>> {
```
Comment (israx) [inline packages/auth/src/Auth.ts:105]: I think this API goes to the Cognito Provider

File: packages/auth/src/Providers/AmazonCognitoProvider.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (israx) [reply packages/auth/src/Providers/AmazonCognitoProvider.ts:92]: Yes, the error handling is still being worked on

File: packages/auth/src/Providers/AmazonCognitoProvider.ts (hunk 17 above)
Comment (israx) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:190]: I think it's not necessary using a generic function here, we can pass directly the **CognitoUserAttributeKey** instead 

```
AuthSignUpResult<CognitoUserAttributeKey>
```

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 20]
```diff
@@ -15,29 +15,320 @@ import {
...
+			await this.signInAfterUserConfirmed(initiateAuthCommand, username);
+		} else if (this._config.signUpVerificationMethod === 'link') {
+			await this.handleLinkAutoSignIn(initiateAuthCommand, username);
+		} else {
+			await this.handleCodeAutoSignIn(initiateAuthCommand, username);
+		}
+
+	}
+
+	private async signInAfterUserConfirmed<UserAttributeKey extends AuthUserAttributeKey>(
```
Comment (israx) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:218]: same as above

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 21]
```diff
@@ -15,29 +15,320 @@ import {
...
+	private handleCodeAutoSignIn(initiateAuthCommand: InitiateAuthCommand, username: string) {
+		const hubListenerCancelToken = Hub.listen('auth', async ({ payload }) => {
+			if (payload.event === 'confirmSignUp') {
+				await this.signInAfterUserConfirmed(initiateAuthCommand, username);
+				hubListenerCancelToken();
+			}
+		});
+	}
+
+	private mapChallengeNames(challengeName: string): AuthSignInStep {
```
Comment (israx) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:292]: I think this method should be part of utils, as it can used by other APIs as well

Comment (israx) [reply packages/auth/src/Providers/AmazonCognitoProvider.ts:292]: And also maybe changing the **challengeName** param type  to **ChallengeNameType**

File: packages/auth/src/types/AmazonCognitoProvider.ts (hunk 18 above)
Comment (israx) [reply packages/auth/src/types/AmazonCognitoProvider.ts:6]: I think we need to follow a similar folder structure for types as the other libraries, iOS is a good point of reference

File: packages/auth/src/types/AmazonCognitoProvider.ts [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+	destination: string;
+};
+
+export enum DeliveryMedium {
+	EMAIL = 'EMAIL',
+	SMS = 'SMS',
+	UNKNOWN = 'UNKNOWN'
+}
+
+export enum AuthProvider {
```
Comment (israx) [inline packages/auth/src/types/AmazonCognitoProvider.ts:39]: This AuthProvider got updated

File: packages/auth/src/types/AmazonCognitoProvider.ts [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+	CONFIRM_SIGN_IN_WITH_SOFTWARE_TOKEN_MFA_CODE = 'CONFIRM_SIGN_IN_WITH_SOFTWARE_TOKEN_MFA_CODE',
+	SELECT_MFA_TYPE = 'SELECT_MFA_TYPE',
+	CONFIRM_SIGN_UP = 'CONFIRM_SIGN_UP',
+	RESET_PASSWORD = 'RESET_PASSWORD',  
+	DONE = 'DONE',
+}
+
+// sign out
+
+export type SignOutResult = {
```
Comment (israx) [inline packages/auth/src/types/AmazonCognitoProvider.ts:172]: Now the the SignOut result has a better type definition

File: packages/auth/src/types/AmazonCognitoProvider.ts [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+
+};
+
+
+// reset password
+
+export type ResetPasswordResult = {
```
Comment (israx) [inline packages/auth/src/types/AmazonCognitoProvider.ts:179]: same as above

File: packages/auth/src/types/hubEvents.ts [hunk 25]
```diff
@@ -0,0 +1,121 @@
+import { AuthError } from '../Errors';
+import { AuthSignInResult, AuthSignUpResult, AuthUserAttributeKey, ResetPasswordResult, SignOutResult } from './AmazonCognitoProvider';
+
+
+export const enum AuthHubEvent {
```
Comment (israx) [inline packages/auth/src/types/hubEvents.ts:5]: Not sure exactly were this should be located at. @haverchuck Do you think it should be part of the category or cognito specific ?

File: packages/auth/src/Providers/AmazonCognitoProvider.ts [hunk 26]
```diff
@@ -15,29 +15,320 @@ import {
...
 	constructor(config?) {
 		this._config = config ? config : {};
+		this._storage = this._config.storage;
+		this.client = new CognitoIdentityProviderClient({ region: config.region });
 	}
-	signUp(): Promise<any> {
-		throw new Error('Method not implemented.');
+
+	async signUp<PluginOptions extends AuthPluginOptions = CognitoSignUpOptions>(req: SignUpRequest<CognitoUserAttributeKey, PluginOptions>): Promise<AuthSignUpResult<CognitoUserAttributeKey>> {
+		if (!this._config?.userPoolId) {
```
Comment (israx) [inline packages/auth/src/Providers/AmazonCognitoProvider.ts:83]: I think here's a good place to define a generic assertion guard to check for plugin options

File: package.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply package.json:126]: Should this be in the auth package's package.json?

**Pull request closed without merging**